File: dynamic_programming/minimizing_coins/minimizing_coins.cpp
#include <iostream>
#include <algorithm>
#include <vector>

#define MAX 90000000000

using namespace std;

int minimizing_coins(int num_coins, vector<int> coins, int sum_size) {
    vector<long long> dp(sum_size + 1, MAX);
    dp[0] = 0;
    for (int i = 1; i < sum_size + 1; i++) {
        for (int j = 0; j < num_coins; j++) {
            int coin = coins[j];
            if (i - coin >= 0) {
                dp[i] =  min(dp[i], dp[i - coin] + 1);
            }
        }
    }
    return dp[sum_size] == MAX ? -1 : dp[sum_size];
}

int main() {
    int sum, n;
    vector<int> coins(0);
    scanf("%d %d", &n, &sum);
    int i = 0;
    int coin;
    while (i < n) {
        scanf("%d", &coin);
        coins.push_back(coin);
        i++;
    }
    cout << minimizing_coins(n, coins, sum) << endl;
}

File: dynamic_programming/minimizing_coins/minimizing_coins.py
from sys import stdin

# TLE use cpp
def minimizing_coins(coins, s):
    dp = [-1 for _ in range(s + 1)]
    dp[0] = 0
    for i in range(1, len(dp)):
        for c in coins:
            if i - c >= 0 and dp[i - c] != -1:
                dp[i] = min(dp[i], dp[i - c] + 1) if dp[i] != -1 else dp[i - c] + 1
    return dp[s]
    
def read_num():
    return int(stdin.readline())

def read_nums():
    return [int(num) for num in stdin.readline().split()]

if __name__ == '__main__':
    _, sum_size = read_nums()
    coins = read_nums()
    print(minimizing_coins(coins, sum_size))
    